backend/src/game.rs

```diff
@@ -69,21 +69,50 @@ impl Game {
 
     #[instrument(skip(self))]
     pub fn move_unit(&mut self, from: Point, to: Point) -> Result<Vec<Point>> {
-        match self.get_unit(from.x, from.y).wrap_err("get unit")? {
-            Some(unit) => {
-                if unit.movements == 0 {
-                    return Err(GameError::NoMoves).wrap_err("no moves left")?;
-                }
-                let mut unit_mut = unit;
-                unit_mut.change_movements(unit.speed);
-                self.set_unit(to.x, to.y, Some(unit_mut))
-                    .wrap_err("failed to move unit; failed to set unit")?;
-                self.set_unit(from.x, from.y, None)
-                    .wrap_err("failed to move unit; failed to unset unit")?;
-                Ok(vec![from, to])
-            }
-            None => Err(GameError::NoUnit).wrap_err("failed to get unit on move")?,
+        let from_hex = match self.get_hex(from.x, from.y) {
+            Some(hex) => hex,
+            None => Err(GameError::NoHex).wrap_err("no hex to move unit from")?,
+        };
+
+        let unit = match from_hex.get_unit() {
+            Some(unit) => unit,
+            None => Err(GameError::NoUnit).wrap_err("no unit on move")?,
+        };
+
+        if unit.movements == 0 {
+            return Err(GameError::NoMoves).wrap_err("no moves left")?;
         }
+
+        let mut hexmap: HashMap<Point, u32> = HashMap::with_capacity(self.field.hexes.len());
+
+        self.path_hexmap(&from_hex, 0, unit.movements, &mut hexmap);
+
+        let path = match self.restore_path_from_hexmap(from, to, &hexmap) {
+            Ok(path) => path,
+            Err(e) => return Err(e.wrap_err("restore path")),
+        };
+
+        let from_hex = match self.get_hex_mut(from.x, from.y) {
+            Some(hex) => hex,
+            None => Err(GameError::NoHex).wrap_err("hex from disappeared after pathfinding")?,
```

Comment:
Is this even possible?

Reply:
In current implementation no

But it should be here if something will be changed in future

Reply:
It probably should be an assert then. Because it's not a game logic error, it's an internal problem.

Reply:
I think about it as more global problem

We have two kinds of mistakes: like this, when something completely wrong and and errors when user click to wrong hex, which is usual situation and we want just pass information to user

I think i should make ticket about this

Reply:
I do not like idea of asserts in the code, because they terminate whole application which is kinda good for testing and okay for mvp
But i think we can do better by creating two groups of errors and make special logic for each kind

Reply:
here is the ticket for further discussions #143